syntax = "proto3";

package moa; // message oriented architecture

service Broker {
  // Publisher actions
  rpc Register(RegisterRequest) returns (Empty) {}
  rpc Publish(stream Message)   returns (Empty) {}
  
  // Subscriber actions
  rpc Subscribe(SubscribeRequest) returns (stream Message) {}
  rpc RequestTags(Empty)          returns (TagsReply)      {}
}

message Empty {}

message RegisterRequest {
  string tag = 1; 
}

message SubscribeRequest {
  string tag = 1; 
}

message TagsReply {
  string list = 1;
}

message Message {
  int32  id        = 1;
  int64  timestamp = 2;
  string tag       = 3;
  string msg       = 4;
}
